File: src/main/java/com/neuronrobotics/bowlerbuilder/BowlerBuilder.java
package com.neuronrobotics.bowlerbuilder;

import com.google.common.base.Throwables;
import com.google.inject.Guice;
import com.google.inject.Injector;
import com.google.inject.Singleton;
import com.neuronrobotics.bowlerbuilder.controller.MainWindowController;
import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;
import javafx.application.Application;
import javafx.fxml.FXMLLoader;
import javafx.scene.Scene;
import javafx.scene.layout.Pane;
import javafx.stage.Stage;

@Singleton
public class BowlerBuilder extends Application {

  private static final Logger LOGGER =
      LoggerUtilities.getLogger(BowlerBuilder.class.getSimpleName());
  private static final Injector INJECTOR = Guice.createInjector();

  @Override
  public void start(final Stage primaryStage) throws IOException {
    //Log uncaught exceptions on the FX thread
    Thread.currentThread().setUncaughtExceptionHandler((exceptionThread, exception) ->
        LOGGER.log(Level.SEVERE, Throwables.getStackTraceAsString(exception)));

    Thread.setDefaultUncaughtExceptionHandler((exceptionThread, exception) ->
        LOGGER.log(Level.SEVERE, Throwables.getStackTraceAsString(exception)));

    final FXMLLoader loader = new FXMLLoader(
        BowlerBuilder.class.getResource("/com/neuronrobotics/bowlerbuilder/MainWindow.fxml"),
        null,
        null,
        INJECTOR::getInstance);
    final Pane mainWindow = loader.load();

    primaryStage.setTitle("BowlerBuilder");
    primaryStage.setScene(new Scene(mainWindow));
    primaryStage.setOnCloseRequest(event ->
        ((MainWindowController) loader.getController()).saveAndQuit());
    primaryStage.show();
  }

  public static Injector getInjector() {
    return INJECTOR;
  }

  public static MainWindowController getMainController() {
    return INJECTOR.getInstance(MainWindowController.class);
  }

}

File: BowlerBuilder/JavaUI/src/main/java/com/neuronrobotics/bowlerbuilder/BowlerBuilder.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.neuronrobotics.bowlerbuilder

import com.google.common.base.Throwables
import com.google.inject.Guice
import com.google.inject.Injector
import com.google.inject.Singleton
import com.neuronrobotics.bowlerbuilder.controller.MainWindowController
import javafx.application.Application
import javafx.fxml.FXMLLoader
import javafx.scene.Scene
import javafx.scene.control.SplitPane
import javafx.stage.Stage
import java.io.IOException
import java.util.logging.Level

@Singleton
class BowlerBuilder : Application() {

    @Throws(IOException::class)
    override fun start(primaryStage: Stage) {
        // Log uncaught exceptions on the FX thread
        Thread.currentThread()
            .setUncaughtExceptionHandler { _, exception ->
                LOGGER.log(Level.SEVERE, Throwables.getStackTraceAsString(exception))
            }

        Thread.setDefaultUncaughtExceptionHandler { _, exception ->
            LOGGER.log(Level.SEVERE, Throwables.getStackTraceAsString(exception))
        }

        val loader = FXMLLoader(
            BowlerBuilder::class.java.getResource("/com/neuronrobotics/bowlerbuilder/MainWindow.fxml"),
            null,
            null,
            { injector.getInstance(it) })

        val mainWindow = loader.load<SplitPane>()

        primaryStage.title = "BowlerBuilder"
        primaryStage.scene = Scene(mainWindow)
        primaryStage.setOnCloseRequest { (loader.getController<Any>() as MainWindowController).saveAndQuit() }
        primaryStage.show()
    }

    companion object {

        private val LOGGER = LoggerUtilities.getLogger(BowlerBuilder::class.java.simpleName)

        @PublicAPI
        @JvmStatic
        val injector: Injector = Guice.createInjector()

        @Suppress("unused")
        @PublicAPI
        @JvmStatic
        val mainController: MainWindowController
            get() = injector.getInstance(MainWindowController::class.java)
    }
}
